[Fix pull-local-node-e2e] Install CNI and Enable containerd for kubekins docker

adisky opened the pull request:
Signed-off-by: Aditi Sharma <adi.sky17@gmail.com>
/sig node
/sig testing
/cc @aojea @dims 

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind failing-test

#### What this PR does / why we need it:
Add support to install CNI when running in CI, also enables cri for docker so Kubelet can connect to docker installed containerd

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Ref https://github.com/kubernetes/test-infra/issues/24715

#### Special notes for your reviewer:
This PR alone is not sufficient to fix job, we need to set some env variables 
test-infra PR :- https://github.com/kubernetes/test-infra/pull/25352

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
None
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

aojea commented on hack/local-up-cluster.sh:
```diff
@@ -1044,6 +1044,93 @@ if [[ "${KUBETEST_IN_DOCKER:-}" == "true" ]]; then
+  CNI_CONFIG_DIR=${DESTDIR}/etc/cni/net.d
+
+  PDIR=$(pwd)
+  cd /workspace 
+  TMPROOT=$(mktemp -d)
+  git clone https://github.com/containernetworking/plugins.git "${TMPROOT}"/plugins
+  pushd "${TMPROOT}"/plugins
+  git checkout v1.0.1
+  ./build_linux.sh
+  mkdir -p $CNI_DIR
+  cp -r ./bin $CNI_DIR
+  mkdir -p $CNI_CONFIG_DIR
+  cat << EOF | tee $CNI_CONFIG_DIR/10-containerd-net.conflist
```
we use this in kind to download the binaries and use just the ones we need

https://github.com/kubernetes-sigs/kind/blob/80a6a355a2065823bbdae5048abf76a4f2a3ec9d/images/base/Dockerfile#L171-L188

int his case, you just need to replace `ptp` by `bridge` in the script

adisky replied:
Thanks, this is really rough script 🙈 , will refine later once it starts passing

aojea commented on hack/local-up-cluster.sh:
```diff
@@ -1044,6 +1044,93 @@ if [[ "${KUBETEST_IN_DOCKER:-}" == "true" ]]; then
+cd $PDIR
+echo "Enabling IPV6 for Docker."
+# configure the daemon with ipv6
+mkdir -p /etc/docker/
+    cat <<EOF >/etc/docker/daemon.json
+{
+  "ipv6": true,
+  "fixed-cidr-v6": "fc00:db8:1::/64"
+}
+EOF
```
why is docker required? is not about containerd only?

adisky replied:
we need docker to run kubetest

adisky committed:
Add support to install cni and enable cri

to run pull-kubernetes-local-e2e on containerd
we need to install cni and enable cri for docker
installed in kubekins image

adisky commented:
@aojea can you take a look now?

aojea commented on hack/local-up-cluster.sh:
```diff
@@ -1044,6 +1112,20 @@ if [[ "${KUBETEST_IN_DOCKER:-}" == "true" ]]; then
+  # install cni for docker in docker
+  # containerd 1.4.12 installed by docker supports CNI version 0.4.0
+  CNI_VERSION=0.4.0
+  install_cni "${CNI_VERSION}"
```
this should be optional to avoid conflicts with current installations, right?

aojea replied:
or just check that `CNI_CONFIG_DIR` is empty

aojea replied:
yeah, definitively I don't think is good idea downloading things by default,

aojea replied:
nevermind, this is correct, it already in 
`if [[ "${KUBETEST_IN_DOCKER:-}" == "true" ]]; then`

adisky replied:
this is only done for 
```
if [[ "${KUBETEST_IN_DOCKER:-}" == "true" ]];
```
which runs in the CI where containerd or any other cri not installed, so won't conflict with any current installtion

aojea commented on hack/local-up-cluster.sh:
```diff
@@ -1044,6 +1112,20 @@ if [[ "${KUBETEST_IN_DOCKER:-}" == "true" ]]; then
+  # we need to enable cri and install cni
+  # install cni for docker in docker
+  # containerd 1.4.12 installed by docker supports CNI version 0.4.0
+  CNI_VERSION=0.4.0
```
this variable is duplicate and is not really needed, just hardcode 0.4.0 to the config file

>   "cniVersion": "${CNI_VERSION}",
to
>   "cniVersion": "0.4.0",

adisky replied:
actually there is a reason not directly hardcoded to config file, the function `install_cni` is intended to be used by directly by the user of the script, so they can set cni version as per there requirement

I would have hardcoded it to 1.0.0 but the containerd installed by docker in kubekins image does't support it

aojea replied:
this version is only the one that goes in the manifest conflist, is not needed

aojea replied:
you have CNI_VERSION at the beginning to represent the version of the binary

adisky replied:
but what i found is `CNI_VERSION` & `CNI_PLUGINS_VERSION` both are different 
`CNI_PLUGINS_VERSION` is the version of release plugins https://github.com/containernetworking/plugins/releases
`CNI_VERSION` is the spec version

adisky replied:
removed `CNI_VERSION`

adisky commented:
/sig node
/sig testing
/kind failing-test
/priority important-soon

aojea commented:
LGTM but ... can we try to get the job fixed in this PR adding new commits?

now it fails because it has the wrong socket endpoint for the runtime

> I0222 04:26:41.859392  141426 remote_runtime.go:71] "Connecting to runtime service" endpoint="unix:///run/containerd/containerd.sock"
W0222 04:26:41.859654  141426 clientconn.go:1331] [core] grpc: addrConn.createTransport failed to connect to {/run/containerd/containerd.sock /run/containerd/containerd.sock <nil> 0 <nil>}. Err: connection error: desc = "transport: Error while dialing dial unix /run/containerd/containerd.sock: connect: no such file or directory". Reconnecting...
Error: failed to run Kubelet: unable to determine runtime API version: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial unix /run/containerd/containerd.sock: connect: no such file or directory"
Usage:

aojea commented:
ahh, the problem is not kubelet, is containerd, right?

/lgtm

adisky commented:
@aoeja have you seen this PR https://github.com/kubernetes/test-infra/pull/25352
the job would get fixed when we merge both

adisky commented:
/test pull-kubernetes-local-e2e
